File: infermap/scorers/initialism.py
"""Initialism / abbreviation scorer.

Matches cases where one field name is an abbreviation formed by taking
a non-empty prefix of each token from the other, in order. Examples:

    assay_id          <-> ASSI    (ASS + I)
    confidence_score  <-> CONSC   (CON + SC)
    relationship_type <-> RELATIT (RELATI + T)
    variant_id        <-> VARI    (VAR + I)
    curated_by        <-> CURAB   (CURA + B)

Fires in both directions (source-is-abbrev or target-is-abbrev). Abstains
(returns None) when neither side is a valid prefix-concat of the other —
otherwise this would emit noise on unrelated pairs.

Uses ``canonical_name`` when MapEngine has populated it, so schema-wide
common affixes are already stripped and we only reason about the
semantic tokens (e.g. "assays_" is gone before we see "assay_id").
"""
from __future__ import annotations

import re

from infermap.types import FieldInfo, ScorerResult


_TOKEN_RE = re.compile(r"[A-Za-z][a-z]*|[A-Z]+(?=[A-Z][a-z]|\b)|\d+")


def _tokenize(name: str) -> list[str]:
    """Split a field name into lowercase tokens.

    Handles snake_case, kebab-case, camelCase, and PascalCase. Numbers are
    their own tokens. Returns lowercase strings with no delimiters.
    """
    # Replace common delimiters with spaces, then split camelCase boundaries.
    cleaned = re.sub(r"[_\-. ]+", " ", name.strip())
    tokens: list[str] = []
    for chunk in cleaned.split():
        # Split camelCase: insert space before an uppercase run
        parts = re.findall(r"[A-Z]+(?=[A-Z][a-z])|[A-Z]?[a-z]+|[A-Z]+|\d+", chunk)
        tokens.extend(p.lower() for p in parts if p)
    return tokens


def _is_prefix_concat(target: str, source_tokens: list[str]) -> bool:
    """Is *target* a concatenation of non-empty prefixes of *source_tokens*
    (in order), using every source token exactly once?

    Each source token must contribute at least 1 character. DP search.
    """
    target = target.lower()
    n_src = len(source_tokens)
    n_tgt = len(target)
    if n_src == 0 or n_tgt == 0:
        return False

    # dp[i][j] = can we consume target[0:j] using source_tokens[0:i]?
    dp = [[False] * (n_tgt + 1) for _ in range(n_src + 1)]
    dp[0][0] = True
    for i in range(1, n_src + 1):
        tok = source_tokens[i - 1]
        for j in range(1, n_tgt + 1):
            # Try each prefix length k of this token (>=1)
            for k in range(1, min(len(tok), j) + 1):
                if target[j - k : j] == tok[:k] and dp[i - 1][j - k]:
                    dp[i][j] = True
                    break
    return dp[n_src][n_tgt]


def _score_pair(name_a: str, name_b: str) -> float | None:
    """Return a score in (0, 1] if one side abbreviates the other, else None.

    Higher scores mean more of the longer side's content is preserved in
    the abbreviation (i.e. the abbreviation is not too lossy).
    """
    tok_a = _tokenize(name_a)
    tok_b = _tokenize(name_b)

    # Normalize the candidate "single-token" side: join into one string
    # without delimiters, lowercased.
    joined_a = "".join(tok_a)
    joined_b = "".join(tok_b)
    if not joined_a or not joined_b:
        return None
    # Degenerate case: identical after tokenization (handled elsewhere).
    if joined_a == joined_b:
        return None

    # Try: b is an abbreviation of a's tokens.
    if _is_prefix_concat(joined_b, tok_a):
        long, short = joined_a, joined_b
    elif _is_prefix_concat(joined_a, tok_b):
        long, short = joined_b, joined_a
    else:
        return None

    # Compression ratio. 1.0 = same length (no compression), 0.0 = empty.
    # Map to (0.6, 0.95] so a 50%-compressed abbreviation scores ~0.78.
    ratio = len(short) / len(long)
    return 0.6 + 0.35 * ratio


class InitialismScorer:
    """Scores source/target pairs where one side is a prefix-concat
    abbreviation of the other's tokens.

    Abstains (returns None) on pairs that are not abbreviations, to avoid
    polluting the weighted average with zeros.
    """

    name: str = "InitialismScorer"
    weight: float = 0.75

    def score(self, source: FieldInfo, target: FieldInfo) -> ScorerResult | None:
        src_name = source.canonical_name or source.name
        tgt_name = target.canonical_name or target.name
        score = _score_pair(src_name, tgt_name)
        if score is None:
            return None
        return ScorerResult(
            score=score,
            reasoning=(
                f"Initialism/abbreviation match: '{src_name}' <-> '{tgt_name}' "
                f"(score={score:.3f})"
            ),
        )

File: packages/infermap-js/src/core/scorers/initialism.ts
// Initialism / abbreviation scorer. Mirrors infermap/scorers/initialism.py.
//
// Matches cases where one field name is an abbreviation formed by taking a
// non-empty prefix of each token from the other, in order. Examples:
//
//   assay_id          <-> ASSI    (ASS + I)
//   confidence_score  <-> CONSC   (CON + SC)
//   relationship_type <-> RELATIT (RELATI + T)
//
// Abstains (returns null) on non-abbreviation pairs.

import type { FieldInfo, ScorerResult } from "../types.js";
import { makeScorerResult } from "../types.js";
import type { Scorer } from "./base.js";

/**
 * Split a field name into lowercase tokens. Handles snake_case, kebab-case,
 * camelCase, PascalCase. Numbers are their own tokens.
 */
export function tokenize(name: string): string[] {
  const cleaned = name.trim().replace(/[_\-. ]+/g, " ");
  const tokens: string[] = [];
  // Match camelCase/PascalCase runs: acronym runs before a CamelWord, Word, or all-caps run, or digits.
  const re = /[A-Z]+(?=[A-Z][a-z])|[A-Z]?[a-z]+|[A-Z]+|\d+/g;
  for (const chunk of cleaned.split(/\s+/)) {
    if (!chunk) continue;
    const matches = chunk.match(re);
    if (!matches) continue;
    for (const m of matches) tokens.push(m.toLowerCase());
  }
  return tokens;
}

/**
 * Is *target* a concatenation of non-empty prefixes of *sourceTokens* in
 * order, using every source token exactly once? DP search.
 */
export function isPrefixConcat(target: string, sourceTokens: string[]): boolean {
  const t = target.toLowerCase();
  const nSrc = sourceTokens.length;
  const nTgt = t.length;
  if (nSrc === 0 || nTgt === 0) return false;
  // Guard against pathological inputs — abort if the DP table would be
  // unreasonably large. 200 chars per side is generous for field names.
  if (nTgt > 200 || nSrc > 50) return false;
  // dp[i][j] = can target[0:j] be consumed using sourceTokens[0:i]
  const dp: boolean[][] = Array.from({ length: nSrc + 1 }, () =>
    new Array<boolean>(nTgt + 1).fill(false)
  );
  dp[0]![0] = true;
  for (let i = 1; i <= nSrc; i++) {
    const tok = sourceTokens[i - 1]!;
    for (let j = 1; j <= nTgt; j++) {
      const maxK = Math.min(tok.length, j);
      for (let k = 1; k <= maxK; k++) {
        if (dp[i - 1]![j - k]! && t.slice(j - k, j) === tok.slice(0, k)) {
          dp[i]![j] = true;
          break;
        }
      }
    }
  }
  return dp[nSrc]![nTgt]!;
}

/**
 * Return a score in (0, 1] if one side abbreviates the other, else null.
 */
export function scorePair(nameA: string, nameB: string): number | null {
  const tokA = tokenize(nameA);
  const tokB = tokenize(nameB);
  const joinedA = tokA.join("");
  const joinedB = tokB.join("");
  if (!joinedA || !joinedB) return null;
  if (joinedA === joinedB) return null;

  let long: string;
  let short: string;
  if (isPrefixConcat(joinedB, tokA)) {
    long = joinedA;
    short = joinedB;
  } else if (isPrefixConcat(joinedA, tokB)) {
    long = joinedB;
    short = joinedA;
  } else {
    return null;
  }
  const ratio = short.length / long.length;
  return 0.6 + 0.35 * ratio;
}

export class InitialismScorer implements Scorer {
  readonly name = "InitialismScorer";
  readonly weight = 0.75;

  score(source: FieldInfo, target: FieldInfo): ScorerResult | null {
    const srcName = source.canonicalName ?? source.name;
    const tgtName = target.canonicalName ?? target.name;
    const s = scorePair(srcName, tgtName);
    if (s === null) return null;
    return makeScorerResult(
      s,
      `Initialism/abbreviation match: '${srcName}' <-> '${tgtName}' (score=${s.toFixed(3)})`
    );
  }
}
